cas-server-core/src/main/java/org/jasig/cas/services/JsonServiceRegistryConfigWatcher.java

```diff
@@ -0,0 +1,188 @@
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.jasig.cas.services;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static java.nio.file.StandardWatchEventKinds.*;
+
+/**
+ * This is {@link JsonServiceRegistryConfigWatcher} that watches the json config directory
+ * for changes and promptly attempts to reload the CAS service registry configuration.
+ *
+ * @author Misagh Moayyed
+ * @since 4.1.0
+ */
+class JsonServiceRegistryConfigWatcher implements Runnable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JsonServiceRegistryConfigWatcher.class);
+
+    private final AtomicBoolean running = new AtomicBoolean(false);
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Lock readLock = this.lock.readLock();
+
+    private final WatchService watcher;
+
+    private final JsonServiceRegistryDao serviceRegistryDao;
+
+    /**
+     * Instantiates a new Json service registry config watcher.
+     *
+     * @param serviceRegistryDao the registry to callback
+     */
+    public JsonServiceRegistryConfigWatcher(final JsonServiceRegistryDao serviceRegistryDao) {
+        try {
+            this.serviceRegistryDao = serviceRegistryDao;
+            this.watcher = FileSystems.getDefault().newWatchService();
+            final WatchEvent.Kind[] kinds = (WatchEvent.Kind[])
+                    Arrays.asList(ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY).toArray();
+            this.serviceRegistryDao.getServiceRegistryDirectory().register(this.watcher, kinds);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (running.compareAndSet(false, true)) {
+            while (running.get()) {
+                // wait for key to be signaled
+                WatchKey key = null;
+                try {
+                    key = watcher.take();
+                    handleEvent(key);
+                } catch (final InterruptedException e) {
+                    return;
```

Comment:
We exit the loop in case of exception and no longer watch the directory, don't we? I would have expected the exception to be ignored.